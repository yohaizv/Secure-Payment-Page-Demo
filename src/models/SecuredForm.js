import { types } from 'mobx-state-tree';
import { Input } from './Input';
import _ from 'lodash';

const selectOption = types
    .model("selectOption", {
        label: types.string,
        value: types.string,
    });

const SecuredFormFields = types
    .model("SecuredFormFields", {
        address: types.optional(Input, { name: "address" }),
        country: types.optional(Input, { name: "country" }),
        cardNumber: types.optional(Input, { name: "cardNumber" }),
        month: types.optional(Input, { name: "month" }),
        year: types.optional(Input, { name: "year" }),
    });



export const FormErrorMessages = {
    FieldsValid: "",
    RequiredFieldsFilled: "Please fill in the missing fields",
    FirstPayment: "The form already submitted"
};
export const SecuredForm = types
    .model("SecuredForm", {
        fields: types.optional(SecuredFormFields, {}),
        countriesOptions: types.optional(types.array(selectOption), [])
    })
    .actions(self => ({
        updateFieldValue(name, value) {
            self.fields[name].value = value;
        },
        updateFieldError(name, error) {
            self.fields[name].error = error;
        }
    }))
    .views(self => {
        //Did all the required fields are filled
        const isRequiredFieldsFilled = () => _.every(self.fields, input => !input.required || input.value);

        //Did all the fields are valid
        const isFieldsValid = () => _.every(self.fields, input => !input.error);

        //Is the first payment action
        const IsFirstPayment = () => !sessionStorage.getItem('formSubmitted');

        return ({
            get isFormValid() {
                return isRequiredFieldsFilled() && isFieldsValid() && IsFirstPayment();
            },
            get formErrorMessage() {
                switch (false) {
                    case isFieldsValid():
                        return FormErrorMessages.FieldsValid;
                    case isRequiredFieldsFilled():
                        return FormErrorMessages.RequiredFieldsFilled;
                    case IsFirstPayment():
                        return FormErrorMessages.FieldsValid;
                    default:
                        return ""
                }
            },
            get monthOptions(){
                return _.filter(['01','02','03','04','05','06','07','08','09','10','11','12'],
                month=>!self.year.value ||
                +self.fields.year.value > new Date().getFullYear() ||
                +month >= new Date().getMonth() + 1
                );
            }

        })
    });



export default SecuredForm;