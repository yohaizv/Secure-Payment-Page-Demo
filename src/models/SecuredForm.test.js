import { getSnapshot, onSnapshot } from "mobx-state-tree";
import { SecuredForm, FormErrorMessages } from "./SecuredForm";
import _ from "lodash";

it("Can create a SecuredForm", () => {
  const securedForm = SecuredForm.create();
  const states = [];

  onSnapshot(securedForm, snapshot => {
    states.push(snapshot);
  });

  securedForm.updateFieldValue("address", "Tel Aviv");
  securedForm.updateFieldError("country", true);

  expect(getSnapshot(securedForm)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
});

it("change SecuredForm field value", () => {
  const securedForm = SecuredForm.create();

  securedForm.updateFieldValue("country", "Tel Aviv");
  expect(securedForm.fields.country.value).toBe("Tel Aviv");
});

it("change SecuredForm field error", () => {
  const securedForm = SecuredForm.create();
  expect(securedForm.fields.country.error).toBe(false);
  securedForm.updateFieldError("country", true);
  expect(securedForm.fields.country.error).toBe(true);
});

it("SecuredForm isFormValid", () => {
  const securedForm = SecuredForm.create();
  securedForm.updateFieldValue("address", "Tel Aviv");
  securedForm.updateFieldValue("country", "Israel");
  securedForm.updateFieldValue("cardNumber", "4580 122345");
  securedForm.updateFieldValue("month", "1");
  expect(securedForm.isFormValid).toBe(false);
  securedForm.updateFieldValue("year", "1");
  expect(securedForm.isFormValid).toBe(true);
  securedForm.updateFieldError("country", true);
  expect(securedForm.isFormValid).toBe(false);
});

it("SecuredForm formErrorMessage", () => {
  const securedForm = SecuredForm.create();
  securedForm.updateFieldValue("address", "Tel Aviv");
  expect(securedForm.formErrorMessage).toBe(
    FormErrorMessages.RequiredFieldsFilled
  );
  securedForm.updateFieldError("cardNumber", true);
  expect(securedForm.formErrorMessage).toBe(FormErrorMessages.FieldsValid);
  securedForm.updateFieldError("cardNumber", false);
  securedForm.updateFieldValue("cardNumber", "4580 122345");
  securedForm.updateFieldValue("country", "Israel");
  securedForm.updateFieldValue("month", "1");
  securedForm.updateFieldValue("year", "1");
  expect(securedForm.formErrorMessage).toBe("");
});

it("Month options", () => {
  const securedForm = SecuredForm.create();
  expect(securedForm.monthOptions.length).toBe(12);

  securedForm.updateFieldValue("year", "2018");
  expect(securedForm.fields.year.value).toBe("2018");
  expect(securedForm.monthOptions).toEqual([11,12]);
});

it("Year options", () => {
  const securedForm = SecuredForm.create();
  expect(securedForm.yearOptions.length).toBe(9);
});
