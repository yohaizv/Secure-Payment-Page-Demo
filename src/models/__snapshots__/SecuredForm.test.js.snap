// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can create a SecuredForm 1`] = `
Object {
  "countriesOptions": Array [],
  "fields": Object {
    "address": Object {
      "error": false,
      "errorMessage": "Invalid Input",
      "name": "address",
      "required": true,
      "rulesTypes": Array [],
      "value": "Tel Aviv",
    },
    "cardNumber": Object {
      "error": false,
      "errorMessage": "Invalid Input",
      "name": "cardNumber",
      "required": true,
      "rulesTypes": Array [],
      "value": "",
    },
    "country": Object {
      "error": true,
      "errorMessage": "Invalid Input",
      "name": "country",
      "required": true,
      "rulesTypes": Array [],
      "value": "",
    },
    "month": Object {
      "error": false,
      "errorMessage": "Invalid Input",
      "name": "month",
      "required": true,
      "rulesTypes": Array [],
      "value": "",
    },
    "year": Object {
      "error": false,
      "errorMessage": "Invalid Input",
      "name": "year",
      "required": true,
      "rulesTypes": Array [],
      "value": "",
    },
  },
}
`;

exports[`Can create a SecuredForm 2`] = `
Array [
  Object {
    "countriesOptions": Array [],
    "fields": Object {
      "address": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "address",
        "required": true,
        "rulesTypes": Array [],
        "value": "Tel Aviv",
      },
      "cardNumber": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "cardNumber",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
      "country": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "country",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
      "month": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "month",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
      "year": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "year",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
    },
  },
  Object {
    "countriesOptions": Array [],
    "fields": Object {
      "address": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "address",
        "required": true,
        "rulesTypes": Array [],
        "value": "Tel Aviv",
      },
      "cardNumber": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "cardNumber",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
      "country": Object {
        "error": true,
        "errorMessage": "Invalid Input",
        "name": "country",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
      "month": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "month",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
      "year": Object {
        "error": false,
        "errorMessage": "Invalid Input",
        "name": "year",
        "required": true,
        "rulesTypes": Array [],
        "value": "",
      },
    },
  },
]
`;
